### PR TITLE
Card spaces endpoint

### PR DESCRIPTION
**Pull request by jurgenwerk**
Ticket: [CS-2067](https://linear.app/cardstack/issue/CS-2067/endpoint-for-saving-card-space-data)

This PR:
- Adds the POST endpoint to /card-spaces route
- Persists the card space fields (url, name, description, button text, profile image, cover image) into the DB
- Stores data to S3 via a worker

# To discuss

## Images

The card space edit form allows for uploading a profile and a cover image. The current implementation for that is very basic–it only supports saving the image urls, assuming some other part of the app will be responsible for uploading, processing and providing the image url prior to calling this endpoint.

Regarding this, I think it's a good time we start discussing how we wanna approach the image uploading part. The decision will likely be affected by the ability to preview or crop the uploaded image (i.e. seeing the image before hitting Create/Save). There is also a dilemma whether we need an another database table for images, or would storing image urls in card spaces suffice.

1. Upload straight from the browser to storage?
- Upload straight to S3 from the browser using a signed url provided by the hub, get the image url and include it in the payload when creating/editing a card space
- Probably the simplest, but limited or no image post-processing options for resizing etc (can probably use only dimensions, size and extension validations in the client)

2. Upload using the hub as a "proxy" to upload to external services (S3, CloudFlare, Imgix,...)
- New endpoint for uploading
- Gives more control over the uploading part (we can resize/resample the image, save metadata,...)
- Hassan mentioned [Cloudflare's image uploading service](https://developers.cloudflare.com/images/) which comes with processing abilities and a CDN  - looks pretty sweet (if pricing works for us)
- Alternatively, we implement the above ourselves using S3 + CloudFront

3. Uploading images using form data when POST/PUT the card space data?
- Does not support previewing the image, inflexible. Probably not a good idea.

4. Another table for images?
- A new table for images would allow us to save metadata about the images (dimensions, size, cropping, variants...) 
- Will we need this? Or does saving the image's url in the card space table suffice?

## Validations

We should also discuss what kind of validations are suitable for specific fields:

### url 
- Unique
- Valid url 

### name
- Max length 50 (arbitrarily chosen for now)?
- Should we allow this to be anything in unicode? 

### description
- Max length 300 (arbitrarily chosen for now)?
- Should we allow this to be anything in unicode?

### category
- Max length 50 (arbitrarily chosen for now)?
- Alphanumeric text only? 
 
### button text
- Per the design, can only be "Visit this [Space | Business | Creator | Person]" 

TODO:

- [x] Add Card Space short did prefix
- [x] Add additional fields (bio, links, donation. Research whether this is a separate table or it belongs to card space data 🤔)
- [x] Add more validations
- [x] More tests that test validations
- [ ] Decide how to approach uploading

### CODE REVIEW

File: packages/hub/services/validators/card-space.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    if (cardSpace.name?.length > MAX_SHORT_FIELD_LENGTH) {
+      errors['name'] = `Max length is ${MAX_SHORT_FIELD_LENGTH}`;
+    }
+
+    if (cardSpace.category?.length > MAX_SHORT_FIELD_LENGTH) {
+      errors['category'] = `Max length is ${MAX_SHORT_FIELD_LENGTH}`;
+    }
+
+    try {
+      new URL(`https://${cardSpace.url}`);
```
Comment (Aierie) [inline packages/hub/services/validators/card-space.ts:49]: Do we need stricter validation? Eg. url matches `x.card.space`?

Comment (jurgenwerk) [reply packages/hub/services/validators/card-space.ts:49]: Yes, good catch.

**Comment by jurgenwerk** [comment]
@Aierie Thanks for your comments and pointing out the missing things. 

>  wondering if this route is meant to be just for creation or also for editing of Card Spaces

This route is just for creation. Later we will add the edit route as well. It will mostly be the same stuff with the exception of the JSON file in S3, which will have to be replaced with a new file containing new new data.

File: packages/did-resolver/index.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ import * as uuidv4 from 'uuid';
 
 const CURRENT_VERSION = 1;
 
-type CardstackIdentifierType = 'PrepaidCardCustomization' | 'MerchantInfo' | 'SupplierInfo';
+type CardstackIdentifierType = 'PrepaidCardCustomization' | 'MerchantInfo' | 'SupplierInfo' | 'CardSpace';
 
 /* eslint-disable @typescript-eslint/naming-convention */
 const DID_TYPE_TO_SHORT_TYPE = {
```
[the remark this answers is not in the document]
Comment (Aierie) [reply packages/did-resolver/index.ts:13]: I think `cs` works! we could also make `DID_TYPE_TO_SHORT_TYPE` a `Record<CardstackIdentifierType, string>` to let Typescript remind us when we add new types in the future.

File: packages/hub/routes/card-spaces.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  id: string;
+  name: string;
+  url: string;
+  description: string;
+  category: string;
+  buttonText: string;
+  profileImageUrl: string;
+  coverImageUrl: string;
+  ownerAddress: string;
+}
```
[the remark this answers is not in the document]
Comment (Aierie) [reply packages/hub/routes/card-spaces.ts:22]: The images are a tricky though, since they are quite card-specific..

**Comment by jurgenwerk** [comment]
@burieberry thanks for your comments–they are also in line with what Ed said on yesterday's call. 

So the takeaway is that it's ok for now to store this data as a single resource, and make the workflow and the cardspace edit form write to it via API.

It was suggested that there could be a `profiles` table, with JSON fields for `profile`, `bio`, `links`, `donations`, and then store further data in these fields. However now that I'm thinking about this, I'd much prefer having a flat attributes table so that we can benefit from Postgres' field validations as much as possible. 

So, I propose we merge this PR, and then in a later PR add the following fields: `bio_title`, `bio_description`, `links`, `donation_amount` etc, together with the PUT endpoint.

**Comment by lukemelia** [comment]
@jurgenwerk shoot, I didn't intend to merge this. (I meant to merge another PR) It was approved, so I think it is OK, but if you need to revert or whatnot, feel free.

**Pull request merged**